design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,294 @@
+ 
+#### Gateway
+For actualy delivering remote service requests RESTful API will be used - it's an established way to execute requested methods and to
+serialize arguments and return values. For sending HTTP requests [Guzzle](https://github.com/guzzle/guzzle) library will be used - it's
+a stable and well supported library which is capable of sending concurrent requests and allows wide customization.
+Guzzle will be configured to use cUrl (cUrl multi) as a backend. When asynchronous requests are sent via Guzzle it returns a promise which can easily
+be converted to Magento's promise.
+#### Lookup
+The information on how to reach services will be provided in env.php like this:
+```php
+return [
+    'service_base_urls' => [
+        'Magento\\Catalog' => 'http://catalog.mydomain.com/',
+        'Magento\\Checkout' => 'https://checkout.mydomain.com/',
+        'Magento\\Customer' => 'http://192.168.1.1/',
+    ],
+];
+```
+The invoker will look up the information based on provided services' class names.
+ 
+#### Changes to RESTful Web API mechanism
+We need to able so send and receive more information when using RESTful web API for inter-service communication than for regular
+communications so to distinguish regular endpoints and endpoints meant for inter-service communications we will declare them in a
+seperate yet similar xml files - _inter_webapi.xml_. Since we declare services by declaring PHP interfaces there is no reason to provide
+human-readable URLs for endpoints meant for inter-service communications, thus we'll make it possible to omit _url_ attribute of the
+_route_ element in inter_webapi.xml and have it autogenerated for us. This way _the invoker_ will be able to determine URL
+for a service based on class name/method name provided. The _method_ attribute will be dropped because otherwise both nodes
+who are calling remote services and nodes hosting the services will have to have the inter_webapi.xml installed.
+Also caching service call results via
+web server (like for services accessed by GET) may be dangerous because the most fresh results are expected from service calls.
+POST method will always be used when sending HTTP requests to remote services.
+_resources_ child element of the _route_ element will be dropped as well because when calling a service remotely
+we don't need to check privileges again - requests are initiated by our code so we assume they're just. In the end enabling a service
+to be called by other remote services will look like this:
+_inter_webapi.xml_
+```xml
+<?xml version="1.0"?>
+<services xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Webapi:etc/inter_webapi.xsd">
+
+    <!-- Product Service -->
+    <service class="Magento\Catalog\Api\ProductRepositoryInterface" method="save"/>
+</services>
+```
+ 
+We will have to pass authenticated user information to remote services so we'll add _X-Magento-User-Type_ and _X-Magento-User-Id_
```

Comment:
Propose to move this to proposal about [authentication and authorization](https://github.com/magento/architecture/pull/48). Current idea is that authentication will be performed on BFF and each service will perform authorization.